changefeedccl: panic when changing primary key on a table with NOT NULL columns

I received the following panic after changing the primary key on a table with a NOT NULL column:

<details>

```
E180927 17:51:08.713757 17206 util/log/crash_reporting.go:203  [n1] a panic has occurred!
E180927 17:51:09.010747 17206 util/log/crash_reporting.go:477  [n1] Reported as error b4adf5e48d784443aeabd6a2d4305efc
panic: Non-nullable column "test3:val" with no value! Index scanned was "primary" with the index key columns (id) and the values (1) [recovered]
	panic: Non-nullable column "test3:val" with no value! Index scanned was "primary" with the index key columns (id) and the values (1)

goroutine 17206 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc4201cee10, 0x3006460, 0xc426e36f40)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:184 +0x11f
panic(0x265bea0, 0xc425d1f2a0)
	/usr/local/go/src/runtime/panic.go:502 +0x229
github.com/cockroachdb/cockroach/pkg/sql/sqlbase.(*RowFetcher).finalizeRow(0xc425f08900, 0x3006460, 0xc42933f0c0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/rowfetcher.go:1245 +0x9d9
github.com/cockroachdb/cockroach/pkg/sql/sqlbase.(*RowFetcher).NextRow(0xc425f08900, 0x3006460, 0xc42933f0c0, 0x2fe4a20, 0xc427d71ec0, 0x0, 0x0, 0x1, 0xc4288fea80, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/rowfetcher.go:997 +0x183
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.kvsToRows.func1(0x3006460, 0xc42933f0c0, 0xc427bdb600, 0x0, 0x1, 0xc42599ac00, 0x4, 0x4, 0xc42599ac00, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed.go:106 +0x275
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.kvsToRows.func2(0x3006460, 0xc42933f0c0, 0xc42678dc80, 0xc425d1c9f0, 0x0, 0x0, 0xc426959d58)
	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed.go:135 +0x36f
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.emitEntries.func2(0x3006460, 0xc42933f0c0, 0x21, 0xc427d61950, 0x15585311579316b9, 0x0, 0x2fefba0)
	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed.go:195 +0x88
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.(*changeAggregator).tick(0xc42900b180, 0xc424f018c8, 0x2)
	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_processors.go:280 +0x5a
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.(*changeAggregator).Next(0xc42900b180, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_processors.go:261 +0x86
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.(*changeFrontier).Next(0xc4290c1500, 0x3006820, 0xc429964000, 0xc426959d40, 0xc42075a050)
	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_processors.go:489 +0xb3
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.Run(0x3006520, 0xc42678dc80, 0x300fa60, 0xc4290c1500, 0x2fe3720, 0xc4271cdc00)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/base.go:170 +0x35
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ProcessorBase).Run(0xc4290c1500, 0x3006520, 0xc42678dc80, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:731 +0x98
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).StartSync(0xc4289bba40, 0x3006520, 0xc42678dc80, 0x2bffd88, 0xc4289e3880, 0x2fe3360)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:599 +0x191
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc4209c5e00, 0xc427bd65a0, 0x0, 0xc428f57430, 0xc426a73600, 0xc42545f098, 0xc428f57220)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:253 +0x879
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.distChangefeedFlow(0x3006520, 0xc4289448d0, 0x3040f00, 0xc42545f000, 0x55db4052a8f8001, 0xc427d61980, 0xc42552bbf0, 0x21, 0xc427d61950, 0x15585311579316b9, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_dist.go:187 +0xe7b
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.(*changefeedResumer).Resume(0x480ba50, 0x3006520, 0xc4289448d0, 0xc429aebe10, 0x2aa06c0, 0xc42545f000, 0xc427b00f00, 0x3030620, 0xc42032d080)
	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_stmt.go:378 +0x401
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).resume.func1(0x3006520, 0xc4289448d0)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:476 +0x2ee
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1(0xc4201cee10, 0x3006460, 0xc426e36f40, 0xc427497820, 0x1e, 0x0, 0x0, 0xc426e17740)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:324 +0xe6
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:319 +0x133
```

</details>

Repro steps:
- start Kafka
- create a table with a NOT NULL column and an editable primary key:

```sql
CREATE TABLE test3 (
                id INT8 NOT NULL,
                val JSONB NOT NULL,
                CONSTRAINT "primary" PRIMARY KEY (id ASC),
                FAMILY "primary" (id, val)
            )
```

- create a changefeed on the table:

```sql
CREATE CHANGEFEED FOR test3 INTO 'kafka://roland-cdc-demo-0004:9092' WITH UPDATED;
```

- insert some data into the table:

```sql
INSERT INTO test3 VALUES (1, '{"foo": "bar"}'::JSON);
```

- (optional) verify that the insert showed up in Kafka using the console consumer
- update the primary key of an existing record:

```sql
UPDATE test3 SET id = 2 WHERE id = 1;
```

Expected behavior: Kafka emits 2 messages, one with key = 1 and value = NULL, another with key = 2 and the contents of the record
Actual behavior: the panic seen above!

Let me know if you need more logging info but it looks like this reproduces reliably.

--- a/pkg/ccl/changefeedccl/changefeed_test.go
+++ b/pkg/ccl/changefeedccl/changefeed_test.go
@@ -525,6 +525,38 @@ func TestChangefeedComputedColumn(t *testing.T) {
 	t.Run(`enterprise`, enterpriseTest(testFn))
 }
 
+func TestChangefeedUpdatePrimaryKey(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	testFn := func(t *testing.T, db *gosql.DB, f testfeedFactory) {
+		sqlDB := sqlutils.MakeSQLRunner(db)
+		// This NOT NULL column checks a regression when used with UPDATE-ing a
+		// primary key column or with DELETE.
+		sqlDB.Exec(t, `CREATE TABLE foo (a INT PRIMARY KEY, b STRING NOT NULL)`)
+		sqlDB.Exec(t, `INSERT INTO foo VALUES (0, 'bar')`)
+
+		foo := f.Feed(t, `CREATE CHANGEFEED FOR foo`)
+		defer foo.Close(t)
+		assertPayloads(t, foo, []string{
+			`foo: [0]->{"a": 0, "b": "bar"}`,
+		})
+
+		sqlDB.Exec(t, `UPDATE foo SET a = 1`)
+		assertPayloads(t, foo, []string{
+			`foo: [0]->`,
+			`foo: [1]->{"a": 1, "b": "bar"}`,
+		})
+
+		sqlDB.Exec(t, `DELETE FROM foo`)
+		assertPayloads(t, foo, []string{
+			`foo: [1]->`,
+		})
+	}
+
+	t.Run(`sinkless`, sinklessTest(testFn))
+	t.Run(`enterprise`, enterpriseTest(testFn))
+}
+
 func TestChangefeedTruncateRenameDrop(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 

--- a/pkg/sql/sqlbase/rowfetcher.go
+++ b/pkg/sql/sqlbase/rowfetcher.go
@@ -1267,7 +1267,9 @@ func (rf *RowFetcher) finalizeRow() error {
 			return nil
 		}
 		if table.neededCols.Contains(int(table.cols[i].ID)) && table.row[i].IsUnset() {
-			if !table.cols[i].Nullable {
+			// If the row was deleted, we'll be missing any non-primary key
+			// columns, including nullable ones, but this is expected.
+			if !table.cols[i].Nullable && !table.rowIsDeleted {
 				var indexColValues []string
 				for _, idx := range table.indexColIdx {
 					if idx != -1 {

